### DISCUSSION TITLE
Support WebAuthn as 2FA method

### DISCUSSION THREAD

**Discussion by pke** [forwardemail/forwardemail.net#182] [Ideas]
For a website I personally prefer WebAuthn using a FIDO2 USB stick. Its built in major browsers and highly secure while being also very convienient at the same time.

It can work alongside OTP and in a later state support even multiple keys.

**Comment by titanism** [comment]
@pke **We listened to your request** 👂!  You can now set up one or more passkeys (similar to GitHub) with Forward Email 🎉  🚀 

![gif](https://github.com/forwardemail/forwardemail.net/assets/101466223/0101e113-ef2f-4886-aaea-e7f5f2397951)

> Happy with our efforts? Write us a short review at <https://www.trustpilot.com/review/forwardemail.net>.  **Don't forget we also support SMTP, IMAP, POP3, &ndash; and more to come!**

## Step by Step Guide for Passkeys and WebAuthn

1. Go to **My Account &rarr; Security** at https://forwardemail.net/my-account/security

2. Click on the "Add New Passkey" button:

    <img width="1132" alt="Screen Shot 2023-12-13 at 4 18 41 PM" src="https://github.com/forwardemail/forwardemail.net/assets/101466223/f6b7fae0-9191-4962-ab23-35d60be32a0e">

3. Once setup, you can log in with any existing passkey:

    <img width="518" alt="Screen Shot 2023-12-13 at 4 18 02 PM" src="https://github.com/forwardemail/forwardemail.net/assets/101466223/b6b09f3e-9263-4951-9d54-e4235218738b">

4. At anytime you can manage your passkeys at **My Account &rarr; Security** at https://forwardemail.net/my-account/security by clicking on the "Manage Passkeys" button, which will bring up this modal:

    <img width="1170" alt="Screen Shot 2023-12-13 at 5 28 55 PM" src="https://github.com/forwardemail/forwardemail.net/assets/101466223/07037a85-1af0-45ca-9efa-01bac4dc352d">

5. You can also edit passkey nicknames (to help you distinguish which devices are which).  After initial passkey creation, you will be automatically prompted to edit the automatically-created nickname (e.g. "USB", "BLE", etc) based off WebAuthn transports detected.

    <img width="529" alt="Screen Shot 2023-12-13 at 5 27 55 PM" src="https://github.com/forwardemail/forwardemail.net/assets/101466223/487a59c6-80c2-40e1-8128-fea7eab1a68d">

6. You will automatically receive emails whenever new passkeys are added or removed (for security precaution):

   <img width="616" alt="Screen Shot 2023-12-13 at 6 21 08 PM" src="https://github.com/forwardemail/forwardemail.net/assets/101466223/109a938b-9acf-4656-a9d9-54f423ca4fae">
   <img width="623" alt="Screen Shot 2023-12-13 at 6 21 26 PM" src="https://github.com/forwardemail/forwardemail.net/assets/101466223/ae801cbb-86d6-4fd4-96d9-3780a59faf7a">

## Additional Documentation

We have also added a new reference in our FAQ at <https://forwardemail.net/faq#do-you-support-passkeys-and-webauthn> and on our About page at https://forwardemail.net/about for December 2023 history update.

## Commit References

* https://github.com/forwardemail/forwardemail.net/commit/b67dae03765f02d31967333f68140175962107a0
* https://github.com/ladjs/passport/commit/b32a46d09ad9ada42a9d5eed6ddd0747e863a52a
* https://github.com/ladjs/passport/releases/tag/v5.1.0
* https://github.com/ladjs/web/releases/tag/v21.0.12
* https://github.com/ladjs/api/releases/tag/v14.1.11

**Reply by titanism** [reply]
The deploy should finish in < 5 minutes 🚀

**Reply by pke** [reply]
Awesome! Begging for this with my email host for years (Webauthn) without any response.

**Reply by titanism** [reply]
@pke migrate to us for SMTP/IMAP if you haven't already! :rocket:

**Reply by titanism** [reply]
> Awesome! Begging for this with my email host for years (Webauthn) without any response.

@pke yikes, we added this in < 24 hours 🔥

**Comment by titanism** [comment]
Pushed fix <https://github.com/forwardemail/forwardemail.net/commit/8237ed3afe386c9c50d4727c5474fb581cd65c24> for legacy database values.

**Reply by titanism** [reply]
Pushed fix <https://github.com/forwardemail/forwardemail.net/commit/d98b92ca5d326f21f892e31a9a638ae797773e92> for known origin mismatch issue.

**Reply by titanism** [reply]
Just pushed final fixes, and now it should work in production.

Deployment should finish within next 10 mins 🚀 

@pke if you get the chance to test it out let us know!

**Reply by C8opmBM** [reply]
I have issues logging in with my yubikey. 

Key registration process goes smoothly, but when logging in, after inserting the pin and touching the key, I get an error as if the key is not registered.
Bitwarden passkey login works.

I have the same key registered and logging in on github just fine, so there is no issues with it or the FIDO2 setup.
Could you please look into it?

![x](https://github.com/forwardemail/forwardemail.net/assets/60513190/06db3cf9-a718-414f-b5fc-278fad47a6ff)
